v3/src/models/data/formula-fn-registry.ts

```diff
@@ -0,0 +1,157 @@
+      const dataSetName = validArgs[0].value
+      const attrName = validArgs[1].value
+      validArgs[0].value = displayNameMap.dataSet[dataSetName]?.id
+      validArgs[1].value = displayNameMap.dataSet[dataSetName]?.attribute[attrName]
+    },
+    evaluate: (args: MathNode[], mathjs: any, scope: FormulaMathJsScope) => {
+      const dataSetId = evaluateNode(args[0], scope)
+      const attrId = evaluateNode(args[1], scope)
+      const zeroBasedIndex = evaluateNode(args[2], scope) - 1
+      return scope.getDataSet(dataSetId)?.getValueAtIndex(zeroBasedIndex, attrId) || ""
+    }
+  },
+
+  // lookupByKey("dataSetName", "attributeName", "keyAttributeName", "keyAttributeValue" | localKeyAttribute)
+  lookupByKey: {
+    rawArgs: true,
+    isAggregate: false,
+    validateArguments: (args: MathNode[]):
+      [ConstantNode<string>, ConstantNode<string>, ConstantNode<string>, MathNode] => {
+      if (args.length !== 4) {
+        throw new Error("lookupByKey function expects exactly 4 arguments")
+      }
+      if (!isConstantStringNode(args[0]) || !isConstantStringNode(args[1]) || !isConstantStringNode(args[2])) {
+        throw new Error("lookupByKey function expects first three arguments to be strings")
+      }
+      return [args[0], args[1], args[2], args[3]]
+    },
+    getDependency: (args: MathNode[]): ILookupDependency => {
+      const validArgs = fnRegistry.lookupByKey.validateArguments(args)
+      return {
+        type: "lookup",
+        dataSetId: validArgs[0].value,
+        attrId: validArgs[1].value,
+        keyAttrId: validArgs[2].value,
+      }
+    },
+    canonicalize: (args: MathNode[], displayNameMap: DisplayNameMap) => {
+      const validArgs = fnRegistry.lookupByKey.validateArguments(args)
+      const dataSetName = validArgs[0].value
+      const attrName = validArgs[1].value
+      const keyAttrName = validArgs[2].value
+      validArgs[0].value = displayNameMap.dataSet[dataSetName]?.id
+      validArgs[1].value = displayNameMap.dataSet[dataSetName]?.attribute[attrName]
+      validArgs[2].value = displayNameMap.dataSet[dataSetName]?.attribute[keyAttrName]
+    },
+    evaluate: (args: MathNode[], mathjs: any, scope: FormulaMathJsScope) => {
+      const dataSetId = evaluateNode(args[0], scope)
+      const attrId = evaluateNode(args[1], scope)
+      const keyAttrId = evaluateNode(args[2], scope)
+      const keyAttrValue = evaluateNode(args[3], scope)
+
+      const dataSet: IDataSet | undefined = scope.getDataSet(dataSetId)
+      if (!dataSet) {
+        return ""
+      }
+      // TODO: is that the right way to access all the dataset cases?
+      // TODO: Optimize? Sort and use binary search?
+      for (const c of dataSet.cases) {
+        const val = dataSet.getValue(c.__id__, keyAttrId)
+        if (val === keyAttrValue) {
+          return dataSet.getValue(c.__id__, attrId) || ""
+        }
+      }
+      return ""
+    },
+  },
+
+  // mean(expression, filterExpression)
```

Comment:
Note that aggregate functions work on numeric values (like `mean`) ignore not just empty values but non-numeric values as well. Also, if an attribute is typed as numeric, values are coerced to numeric where possible; e.g. the value '1a' is coerced to 1 and included in the calculation. You can see this in V2.

Reply:
Thanks, good to know. Yes, all my PRs totally miss handling of user errors, edge cases, and tests. I didn't try to get into that yet.